src/c++/perf_analyzer/memory_manager.cc

```diff
@@ -0,0 +1,656 @@
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <algorithm>
+
+#include "memory_manager.h"
+
+#ifdef TRITON_ENABLE_GPU
+#include <cuda_runtime_api.h>
+
+#define RETURN_IF_CUDA_ERR(FUNC)                               \
+  {                                                            \
+    const cudaError_t result = FUNC;                           \
+    if (result != cudaSuccess) {                               \
+      return cb::Error(                                        \
+          "CUDA exception (line " + std::to_string(__LINE__) + \
+              "): " + cudaGetErrorName(result) + " (" +        \
+              cudaGetErrorString(result) + ")",                \
+          pa::GENERIC_ERROR);                                  \
+    }                                                          \
+  }
+
+#endif  // TRITON_ENABLE_GPU
+
+namespace triton { namespace perfanalyzer {
+
+
+MemoryManager::~MemoryManager()
+{
+  cb::Error err;
+  if (using_shared_memory_ && backend_.get() != nullptr) {
+    err = backend_->UnregisterAllSharedMemory();
+    if (!err.IsOk()) {
+      std::cerr << "Unable to unregister all shared memory regions"
+                << std::endl;
+    }
+    if (shared_memory_type_ == SharedMemoryType::SYSTEM_SHARED_MEMORY) {
+      for (auto& region : shared_memory_regions_) {
+        if (factory_->Kind() !=
+            triton::perfanalyzer::clientbackend::BackendKind::TRITON_C_API) {
+          err = backend_->UnmapSharedMemory(
+              shared_memory_regions_[region.first].data_.get(),
+              shared_memory_regions_[region.first].byte_size_);
+          if (!err.IsOk()) {
+            std::cerr << "Unable to unmap shared memory with key ("
+                      << region.first << "): Starting: "
+                      << static_cast<void*>(
+                             shared_memory_regions_[region.first].data_.get())
+                      << ", size: "
+                      << shared_memory_regions_[region.first].byte_size_
+                      << std::endl;
+          }
+          err = backend_->UnlinkSharedMemoryRegion(region.first);
+          if (!err.IsOk()) {
+            std::cerr << "Unable to unlink shared memory with key: "
+                      << region.first << std::endl;
+          }
+        }
+      }
+    }
+  }
+}
+
+MemoryManager::MemoryManager(
+    const int32_t batch_size, const SharedMemoryType shared_memory_type,
+    const size_t output_shm_size, const std::shared_ptr<ModelParser>& parser,
+    const std::shared_ptr<cb::ClientBackendFactory>& factory,
```

Comment:
I'm struggling to believe that we need to pass in the full factory. If factory->Kind is all that is used, then pass that in instead.

Reply:
I was wrong. I thought it was only using factory_->Kind(), but it is also using factory_->CreateClientBackend(), which means it needs the full factory